perfmetrics/scripts/load_tests/python/load_generator/load_generator.py

```diff
@@ -0,0 +1,352 @@
+  class also provides default implementation of post load test task to get 
+  latencies of tasks. Classes derived from this class can define
+  their own pre- and post-load test tasks and run before and after actual load
+  test.
+
+  Args:
+    num_processes: An integer defining the number of processes to run during
+      load test.
+    num_threads_per_process: An integer defining the number of threads to run
+      in each process during load test.
+    run_time: An integer defining the number of seconds to run the load test
+      for.
+    num_tasks_per_thread: An integer defining the number of times the given
+      task to be run inside each thread of each process during load test.
+  """
+
+  def __init__(self,
+               num_processes,
+               num_threads_per_process,
+               run_time=sys.maxsize,
+               num_tasks_per_thread=sys.maxsize):
+    self.num_processes = num_processes
+    self.num_threads_per_process = num_threads_per_process
+    self.run_time = min(sys.maxsize, run_time)
+    self.num_tasks_per_thread = min(sys.maxsize, num_tasks_per_thread)
+
+    if (self.run_time == sys.maxsize) & \
+        (self.num_tasks_per_thread == sys.maxsize):
+      raise ValueError('Out of run_time and num_tasks_per_thread '
+                       'arguments, one has to be passed.')
+
+    self.total_num_tasks = sys.maxsize
+    if self.num_tasks_per_thread != sys.maxsize:
+      self.total_num_tasks = self.num_tasks_per_thread * \
+                             self.num_threads_per_process * self.num_processes
+
+  def pre_load_test(self):
+    """Task to perform before running load test.
+    """
+    pass
+
+  def generate_load(self, task):
+    """Generates load with the given task.
+
+    The load is generated on CPU and other resources (depending upon the task
+    used) by running process(s) and thread(s) where task runs inside thread(s)
+    and thread(s) runs inside process(s).
+
+    Args:
+      task: Implementation of task.LoadTestTask.
+
+    Returns:
+      Returns start_time, end_time of load test, latencies and results of all
+        the tasks performed over the span of load test.
+
+    Raises:
+      RuntimeError: If the given task is not completed even once during the
+        course of load test.
+    """
+    tasks_results_queue = multiprocessing.Manager().Queue()
+    pre_tasks_results_queue = multiprocessing.Manager().Queue()
+    post_tasks_results_queue = multiprocessing.Manager().Queue()
+
+    processes = []
+    for process_id in range(self.num_processes):
+      process = multiprocessing.Process(
+          target=LoadGenerator._process_task,
+          args=(task, process_id, self.num_threads_per_process,
+                self.num_tasks_per_thread, pre_tasks_results_queue,
+                tasks_results_queue, post_tasks_results_queue))
+      processes.append(process)
+
+    # Checkpoints initialisation. Note: Only used if self.run_time is set.
```

Comment:
can you elaborate this comment, I am not able to follow

Reply:
Updated the comment: 
"Initialize checkpoints to show completion of load test i.e. 25%, 50% etc.
 Note: Completion checkpoints are shown only when self.run_time is set.
 E.g. if self.run_time is 60 then the load test will inform that 50% of
 load test is completed after 30 seconds."